### PR TITLE
[toolformer] replace "residents" with "people"

### PR DESCRIPTION
**Pull request by chejennifer**
example: https://screenshot.googleplex.com/C3QjzRnnzD9euVc

sv index differ: https://storage.mtls.cloud.google.com/datcom-embedding-diffs/chejennifer_base_uae_mem_2024_07_08_12_11_23.html where test replaces residents with people

### CODE REVIEW

File: server/lib/nl/detection/utils.py [hunk 1]
```diff
@@ -309,3 +309,12 @@ def find_word_boundary(haystack: str, needle: str):
...
     return match.start()
   return -1
+
+
+# Replaces strings in a query given a dictionary where key is the original
+# string and value is the replacement string to use
+def replace_strings_in_query(query: str, replacements: Dict[str, str]) -> str:
+  processed_query = query
+  for orig, new in replacements.items():
+    processed_query = processed_query.replace(orig, new)
```
[the remark this answers is not in the document]
Comment (chejennifer) [reply server/lib/nl/detection/utils.py]: done

**Comment by chejennifer** [comment]
> Thanks!
> 
> How does the svindex_differ look? If needed its useful to disable the "toolformer only" check to get the NL diffs and svindex differ. So we make this change knowing what the losses are (if any)?

There were no golden diffs & these are the svindex diffs: https://storage.mtls.cloud.google.com/datcom-embedding-diffs/chejennifer_base_uae_mem_2024_07_08_12_11_23.html 

- diff 1: might be a concern because svs like Count_Person_25OrMoreYears_BachelorsDegreeOrHigher_AsFractionOf_Count_Person_25OrMoreYears now have a score over 0.8 even though the query asks about "college degree" and I'm not sure "bachelors degree and higher" encompasses all college degrees
- diff 2: no concern
- diff 3: might be a concern because there are more svs with a score over 0.8, but top 2 svs are the same as before and had a score of over 0.8 before as well

File: server/lib/nl/detection/utils.py [hunk 2]
```diff
@@ -309,3 +309,12 @@ def find_word_boundary(haystack: str, needle: str):
...
     return match.start()
   return -1
+
+
+# Replaces strings in a query given a dictionary where key is the original
+# string and value is the replacement string to use
+def replace_strings_in_query(query: str, replacements: Dict[str, str]) -> str:
+  processed_query = query
+  for orig, new in replacements.items():
+    processed_query = re.sub(rf"\b{orig}\b", new, processed_query)
```
Comment (pradh) [inline server/lib/nl/detection/utils.py:319]: nit: comment - `\b` is for ensuring replacements happen in word boundary ?

**Comment by pradh** [comment]
> > Thanks!
> > How does the svindex_differ look? If needed its useful to disable the "toolformer only" check to get the NL diffs and svindex differ. So we make this change knowing what the losses are (if any)?
> 
> There were no golden diffs & these are the svindex diffs: https://storage.mtls.cloud.google.com/datcom-embedding-diffs/chejennifer_base_uae_mem_2024_07_08_12_11_23.html
> 
> * diff 1: might be a concern because svs like Count_Person_25OrMoreYears_BachelorsDegreeOrHigher_AsFractionOf_Count_Person_25OrMoreYears now have a score over 0.8 even though the query asks about "college degree" and I'm not sure "bachelors degree and higher" encompasses all college degrees

I seems {Associate, Bachelor's, Master's, Doctoral} are all different college degrees levels.  So the right answer would be the topic, which anyway doesn't work for RIG.  Perhaps something to consider is making `dc/topic/CollegeEducationAttainment` be the top result for this query, but that doesn't impact RIG, so maybe add a bug/TODO to improve that?

> * diff 2: no concern
> * diff 3: might be a concern because there are more svs with a score over 0.8, but top 2 svs are the same as before and had a score of over 0.8 before as well

Only the first variable seems relevant, even the 2nd one was already irrelevant.  So maybe ok?

**Review by pradh — APPROVED** [review]
Thanks for the changes!

**Comment by chejennifer** [comment]
> > > Thanks!
> > > How does the svindex_differ look? If needed its useful to disable the "toolformer only" check to get the NL diffs and svindex differ. So we make this change knowing what the losses are (if any)?
> > 
> > 
> > There were no golden diffs & these are the svindex diffs: https://storage.mtls.cloud.google.com/datcom-embedding-diffs/chejennifer_base_uae_mem_2024_07_08_12_11_23.html
> > 
> > * diff 1: might be a concern because svs like Count_Person_25OrMoreYears_BachelorsDegreeOrHigher_AsFractionOf_Count_Person_25OrMoreYears now have a score over 0.8 even though the query asks about "college degree" and I'm not sure "bachelors degree and higher" encompasses all college degrees
> 
> I seems {Associate, Bachelor's, Master's, Doctoral} are all different college degrees levels. So the right answer would be the topic, which anyway doesn't work for RIG. Perhaps something to consider is making `dc/topic/CollegeEducationAttainment` be the top result for this query, but that doesn't impact RIG, so maybe add a bug/TODO to improve that?
> 
> > * diff 2: no concern
> > * diff 3: might be a concern because there are more svs with a score over 0.8, but top 2 svs are the same as before and had a score of over 0.8 before as well
> 
> Only the first variable seems relevant, even the 2nd one was already irrelevant. So maybe ok?

For the first diff, the impact on RIG would be the fact that these individual college levels used to have a score less than 0.8 so wouldn't be returned, but now have a score greater than 0.8 so will be returned. But I guess we can also try using LLM to reject the answer here

File: server/lib/nl/detection/utils.py (hunk 2 above)
Comment (chejennifer) [reply server/lib/nl/detection/utils.py:321]: done

**Comment by pradh** [comment]
> > > > Thanks!
> > > > How does the svindex_differ look? If needed its useful to disable the "toolformer only" check to get the NL diffs and svindex differ. So we make this change knowing what the losses are (if any)?
> > > 
> > > 
> > > There were no golden diffs & these are the svindex diffs: https://storage.mtls.cloud.google.com/datcom-embedding-diffs/chejennifer_base_uae_mem_2024_07_08_12_11_23.html
> > > 
> > > * diff 1: might be a concern because svs like Count_Person_25OrMoreYears_BachelorsDegreeOrHigher_AsFractionOf_Count_Person_25OrMoreYears now have a score over 0.8 even though the query asks about "college degree" and I'm not sure "bachelors degree and higher" encompasses all college degrees
> > 
> > 
> > I seems {Associate, Bachelor's, Master's, Doctoral} are all different college degrees levels. So the right answer would be the topic, which anyway doesn't work for RIG. Perhaps something to consider is making `dc/topic/CollegeEducationAttainment` be the top result for this query, but that doesn't impact RIG, so maybe add a bug/TODO to improve that?
> > > * diff 2: no concern
> > > * diff 3: might be a concern because there are more svs with a score over 0.8, but top 2 svs are the same as before and had a score of over 0.8 before as well
> > 
> > 
> > Only the first variable seems relevant, even the 2nd one was already irrelevant. So maybe ok?
> 
> For the first diff, the impact on RIG would be the fact that these individual college levels used to have a score less than 0.8 so wouldn't be returned, but now have a score greater than 0.8 so will be returned. But I guess we can also try using LLM to reject the answer here

Ohh, I didn't realize LHS had no variable above 0.8.  We'll need to try out if LLM rejects, it might not.

**Comment by pradh** [comment]
> > > > > Thanks!
> > > > > How does the svindex_differ look? If needed its useful to disable the "toolformer only" check to get the NL diffs and svindex differ. So we make this change knowing what the losses are (if any)?
> > > > 
> > > > 
> > > > There were no golden diffs & these are the svindex diffs: https://storage.mtls.cloud.google.com/datcom-embedding-diffs/chejennifer_base_uae_mem_2024_07_08_12_11_23.html
> > > > 
> > > > * diff 1: might be a concern because svs like Count_Person_25OrMoreYears_BachelorsDegreeOrHigher_AsFractionOf_Count_Person_25OrMoreYears now have a score over 0.8 even though the query asks about "college degree" and I'm not sure "bachelors degree and higher" encompasses all college degrees
> > > 
> > > 
> > > I seems {Associate, Bachelor's, Master's, Doctoral} are all different college degrees levels. So the right answer would be the topic, which anyway doesn't work for RIG. Perhaps something to consider is making `dc/topic/CollegeEducationAttainment` be the top result for this query, but that doesn't impact RIG, so maybe add a bug/TODO to improve that?
> > > > * diff 2: no concern
> > > > * diff 3: might be a concern because there are more svs with a score over 0.8, but top 2 svs are the same as before and had a score of over 0.8 before as well
> > > 
> > > 
> > > Only the first variable seems relevant, even the 2nd one was already irrelevant. So maybe ok?
> > 
> > 
> > For the first diff, the impact on RIG would be the fact that these individual college levels used to have a score less than 0.8 so wouldn't be returned, but now have a score greater than 0.8 so will be returned. But I guess we can also try using LLM to reject the answer here
> 
> Ohh, I didn't realize LHS had no variable above 0.8. We'll need to try out if LLM rejects, it might not.

https://colab.corp.google.com/google_src/cloud/shanth/wikipedia2/google3/experimental/users/shanth/data_gemma/eval/DC_QA_Validation.ipynb

**Comment by chejennifer** [comment]
> > > > > > Thanks!
> > > > > > How does the svindex_differ look? If needed its useful to disable the "toolformer only" check to get the NL diffs and svindex differ. So we make this change knowing what the losses are (if any)?
> > > > > 
> > > > > 
> > > > > There were no golden diffs & these are the svindex diffs: https://storage.mtls.cloud.google.com/datcom-embedding-diffs/chejennifer_base_uae_mem_2024_07_08_12_11_23.html
> > > > > 
> > > > > * diff 1: might be a concern because svs like Count_Person_25OrMoreYears_BachelorsDegreeOrHigher_AsFractionOf_Count_Person_25OrMoreYears now have a score over 0.8 even though the query asks about "college degree" and I'm not sure "bachelors degree and higher" encompasses all college degrees
> > > > 
> > > > 
> > > > I seems {Associate, Bachelor's, Master's, Doctoral} are all different college degrees levels. So the right answer would be the topic, which anyway doesn't work for RIG. Perhaps something to consider is making `dc/topic/CollegeEducationAttainment` be the top result for this query, but that doesn't impact RIG, so maybe add a bug/TODO to improve that?
> > > > > * diff 2: no concern
> > > > > * diff 3: might be a concern because there are more svs with a score over 0.8, but top 2 svs are the same as before and had a score of over 0.8 before as well
> > > > 
> > > > 
> > > > Only the first variable seems relevant, even the 2nd one was already irrelevant. So maybe ok?
> > > 
> > > 
> > > For the first diff, the impact on RIG would be the fact that these individual college levels used to have a score less than 0.8 so wouldn't be returned, but now have a score greater than 0.8 so will be returned. But I guess we can also try using LLM to reject the answer here
> > 
> > 
> > Ohh, I didn't realize LHS had no variable above 0.8. We'll need to try out if LLM rejects, it might not.
> 
> https://colab.corp.google.com/google_src/cloud/shanth/wikipedia2/google3/experimental/users/shanth/data_gemma/eval/DC_QA_Validation.ipynb

ah, ok, looks like LLM does not reject: https://screenshot.googleplex.com/5xAmnNMoXcRUeUA

so do we still want this change ..? this isn't a query used in our evals

**Comment by pradh** [comment]
> > > > > > > Thanks!
> > > > > > > How does the svindex_differ look? If needed its useful to disable the "toolformer only" check to get the NL diffs and svindex differ. So we make this change knowing what the losses are (if any)?
> > > > > > 
> > > > > > 
> > > > > > There were no golden diffs & these are the svindex diffs: https://storage.mtls.cloud.google.com/datcom-embedding-diffs/chejennifer_base_uae_mem_2024_07_08_12_11_23.html
> > > > > > 
> > > > > > * diff 1: might be a concern because svs like Count_Person_25OrMoreYears_BachelorsDegreeOrHigher_AsFractionOf_Count_Person_25OrMoreYears now have a score over 0.8 even though the query asks about "college degree" and I'm not sure "bachelors degree and higher" encompasses all college degrees
> > > > > 
> > > > > 
> > > > > I seems {Associate, Bachelor's, Master's, Doctoral} are all different college degrees levels. So the right answer would be the topic, which anyway doesn't work for RIG. Perhaps something to consider is making `dc/topic/CollegeEducationAttainment` be the top result for this query, but that doesn't impact RIG, so maybe add a bug/TODO to improve that?
> > > > > > * diff 2: no concern
> > > > > > * diff 3: might be a concern because there are more svs with a score over 0.8, but top 2 svs are the same as before and had a score of over 0.8 before as well
> > > > > 
> > > > > 
> > > > > Only the first variable seems relevant, even the 2nd one was already irrelevant. So maybe ok?
> > > > 
> > > > 
> > > > For the first diff, the impact on RIG would be the fact that these individual college levels used to have a score less than 0.8 so wouldn't be returned, but now have a score greater than 0.8 so will be returned. But I guess we can also try using LLM to reject the answer here
> > > 
> > > 
> > > Ohh, I didn't realize LHS had no variable above 0.8. We'll need to try out if LLM rejects, it might not.
> > 
> > 
> > https://colab.corp.google.com/google_src/cloud/shanth/wikipedia2/google3/experimental/users/shanth/data_gemma/eval/DC_QA_Validation.ipynb
> 
> ah, ok, looks like LLM does not reject: https://screenshot.googleplex.com/5xAmnNMoXcRUeUA
> 
> so do we still want this change ..? this isn't a query used in our evals

Thought so.  But yes, this change I think is OK.  We can keep an eye on "residents" in the NL query to see how they do in RIG...

**Pull request merged**